create /proc/sys/kernel/spl/gitrev with git hash

ahrens opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
The existing mechanisms for determining what code is running in the kernel do not always correctly report the git hash.  The versions reported there do not reflect changes made since `configure` was run (i.e. incremental builds do not update the version) and they are misleading if git tags are not set up properly.  This applies to `modinfo zfs`, `dmesg`, and `/sys/module/zfs/version`.

This is related to https://github.com/zfsonlinux/zfs/issues/7931.  It doesn't directly fix it (existing version info can still be stale), but it provides an alternative which decreases the need for a fix.

### Description
<!--- Describe your changes in detail -->
There are complicated requirements on how the existing version is generated.  Therefore we are leaving that alone, and adding a new mechanism to record and retrieve the git hash: `cat /proc/sys/kernel/spl/gitrev`

The gitrev is re-generated at compile time, when running `make` (including for incremental builds).  The value is the output of `git describe` (or "unknown" if not in a git repo or there are uncommitted changes).

We're also removing /proc/sys/kernel/spl/version, which was never very useful.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

Check the contents of /proc/sys/kernel/spl/gitrev after full and incremental builds, and when there are uncommitted changes.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).
- [ ] Change has been approved by a ZFS on Linux member.

pzakha commented on scripts/make_gitrev.sh:
```diff
@@ -0,0 +1,24 @@
+#!/bin/sh
+
+file=include/spl/sys/zfs_gitrev.h
```
We should also add this to .gitignore

behlendorf replied:
Let's also add a comment explaining that this isn't done with a `zfs_gitrev.h.in` file since it needs to be generated for every invocation of `make`, not once during `configure`.

pzakha commented on scripts/make_gitrev.sh:
```diff
@@ -0,0 +1,24 @@
+#
+# Check if there are uncommitted changes
+#
+git diff-index --quiet HEAD -- || exit
```
Should we perhaps still generate a hash with "(modified)" appended to it, just so that people aren't too confused why it's suddenly "unknown"?
Also nit: `--` looks unnecessary

ahrens replied:
For simplicity, I'm going to drop this (adding `(modified)`) if that's OK with you.

pzakha replied:
It's up to you. That should be very easy to implement though.

pzakha commented on scripts/make_gitrev.sh:
```diff
@@ -0,0 +1,24 @@
+#
+git diff-index --quiet HEAD -- || exit
+
+rev=$(git describe 2>/dev/null)
```
add `|| exit`

behlendorf commented on module/spl/spl-generic.c:
```diff
@@ -45,9 +45,10 @@
-char spl_version[32] = "SPL v" ZFS_META_VERSION "-" ZFS_META_RELEASE;
-EXPORT_SYMBOL(spl_version);
+char spl_gitrev[64] = ZFS_META_GITREV;
+EXPORT_SYMBOL(spl_gitrev);
```
Let's drop the `EXPORT_SYMBOL`, I don't think there's a need for this to be accessible outside the `spl.ko`.

behlendorf commented on scripts/make_gitrev.sh:
```diff
@@ -0,0 +1,24 @@
+# Set default file contents in case we bail.
+#
+rm -f $file
+echo "#define\tZFS_META_GITREV \"unknown\"" >>$file
```
s/echo/echo -e/

```
       -e     enable interpretation of backslash escapes
```

ahrens replied:
Thanks, that explains the build failures!

behlendorf replied:
The `printf` built in should work too as an alternative.

richardelling replied:
printf isn't a builtin for sh, but it does usually exist as /usr/bin/printf

behlendorf commented on scripts/make_gitrev.sh:
```diff
@@ -0,0 +1,24 @@
+echo "#define\tZFS_META_GITREV \"unknown\"" >>$file
+
+#
+# Check if we are in a git repo.
```
...and the `git` is even installed on the system.  It's possible to be building from a tarball.

behlendorf commented on scripts/make_gitrev.sh:
```diff
@@ -0,0 +1,24 @@
+rev=$(git describe 2>/dev/null)
+
+rm -f $file
+echo "#define\tZFS_META_GITREV \"${rev}\"" >>$file
```
s/echo/echo -e/

dweeezil commented:
How about adding `spl_gitrev` to the string logged in `spa_history_log_version()`?

behlendorf commented on Makefile.am:
```diff
@@ -25,6 +25,12 @@ EXTRA_DIST += META AUTHORS COPYRIGHT LICENSE NEWS NOTICE README.md
+gitrev:
+	-${top_srcdir}/scripts/make_gitrev.sh
+
+BUILT_SOURCES = gitrev
```
I think you should add a call to `-${top_srcdir}/scripts/make_gitrev.sh` in `dist-hook:`.  This target is run at the end of `make dist` and it let's you add additional files prior to the `tar`.  Right now it's used to ensure an updated META file is included in the the final tarball.  This would ensure that tarball also includes the git hash if available, and a valid version of `sys/zfs_gitrev.h`.

ahrens replied:
I added a dependency on `gitrev`, which I think will do the same thing without duplicating code.  Let me know if you see a problem with this (I'm not a makefile expert).

ahrens replied:
Well, it does the same thing as what you suggested.  However, zfs_gitrev.h doesn't end up in the tarball.  I suspect that we would need to tell it to generate the file in an alternate location.

ahrens replied:
Turns out we copy the .c/.h files before doing `dist-hook`.  So I had to also add a line to dist-hook to copy zfs_gitrev.h into the distdir.  Now I have:
```
$ make dist
...
make[2]: Entering directory '/export/home/delphix/zfs'
./scripts/make_gitrev.sh
cp ./include/zfs_gitrev.h zfs-0.8.0/include; \
sed -i 's/Release:[[:print:]]*/Release:      rc1/' \
	zfs-0.8.0/META
...
$ tar -tzf zfs-0.8.0.tar.gz |grep git
zfs-0.8.0/scripts/make_gitrev.sh
zfs-0.8.0/include/zfs_gitrev.h
```

ahrens commented:
> How about adding `spl_gitrev` to the string logged in `spa_history_log_version()`?

Great idea.  What would you think about replacing `"software version 5000/5"` with `software version <gitrev>`? I would argue that the `5000/5` is basically useless.  It's been the same for years despite numerous software changes.

ahrens commented:
@behlendorf I'm getting the following errors, do you know how I should address them?  I think I'm not using POSIX sh's echo, I'm using /bin/echo.  
```
./scripts/make_gitrev.sh:37:11: warning: In POSIX sh, echo flags are undefined. [SC2039]
./scripts/make_gitrev.sh:52:11: warning: In POSIX sh, echo flags are undefined. [SC2039]
Makefile:1536: recipe for target 'shellcheck' failed
```
Incidentally, I don't get these errors when running `make checkstyle` locally, because I don't have `shellcheck` installed.  I don't think it's a good idea to have this silently pass.  We could at least print a warning if `shellcheck` is not installed.

dweeezil commented:
@ahrens Yes, that's exactly what I was thinking.  The existing "software version" thing doesn't seem very useful at all.

ahrens commented:
Regarding the shellcheck errors, I noticed that several other scripts use `echo -e` but they are bash scripts.  Should I change the new script to be a bash script, rather than sh?

ahrens commented:
I also see that the kernel.org build is not working, it seems that it isn't generating the new header or it's being put in the wrong place.
```
fs/zfs/spl/spl-generic.c:48:10: fatal error: zfs_gitrev.h: No such file or directory
 #include "zfs_gitrev.h"
          ^~~~~~~~~~~~~~
compilation terminated.
```

ahrens commented:
I believe I've addressed all the review feedback and this is now ready for final review and integration.  Note that I moved the location of the new (generated) header to `include/zfs_gitrev.h`.  This was needed so that the zfs kernel module could also use it (in spa_history_log_version).
